internal/pkg/scaffold/olm-catalog/package_manifest.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package catalog
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/operator-framework/operator-sdk/internal/pkg/scaffold"
+	"github.com/operator-framework/operator-sdk/internal/pkg/scaffold/input"
+
+	"github.com/ghodss/yaml"
+	olmregistry "github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry"
+	"github.com/pkg/errors"
+	"github.com/spf13/afero"
+)
+
+const PackageManifestPrefix = ".package.yaml"
+
+type PackageManifest struct {
+	input.Input
+
+	// CSVVersion is the version of the CSV being updated.
+	CSVVersion string
+	// Channel is CSVVersion's package manifest channel. If a new package
+	// manifest is generated, this channel will be the manifest default.
+	Channel string
+	// If ChannelIsDefault is true, Channel will be the package manifests'
+	// default channel.
+	ChannelIsDefault bool
+}
+
+var _ input.File = &PackageManifest{}
+
+// GetInput gets s' Input.
+func (s *PackageManifest) GetInput() (input.Input, error) {
+	if s.Path == "" {
+		lowerProjName := strings.ToLower(s.ProjectName)
+		// Path is what the operator-registry expects:
+		// {manifests -> olm-catalog}/{operator_name}/{operator_name}.package.yaml
+		s.Path = filepath.Join(OLMCatalogDir, lowerProjName,
+			lowerProjName+PackageManifestPrefix)
+	}
+	return s.Input, nil
+}
+
+var _ scaffold.CustomRenderer = &PackageManifest{}
+
+// SetFS is a no-op to implement CustomRenderer.
+func (s *PackageManifest) SetFS(_ afero.Fs) {}
+
+// CustomRender either reads an existing package manifest or creates a new
+// manifest and modifies it based on values set in s.
+func (s *PackageManifest) CustomRender() ([]byte, error) {
+	i, err := s.GetInput()
+	if err != nil {
+		return nil, err
+	}
+	path := filepath.Join(s.AbsProjectPath, i.Path)
+
+	pm := s.newPackageManifest()
+	if _, err := os.Stat(path); err != nil && !os.IsNotExist(err) {
+		return nil, errors.Wrapf(err, "package manifest %s", path)
+	} else if err == nil {
+		b, err := ioutil.ReadFile(path)
+		if err != nil {
+			return nil, errors.Wrapf(err, "read package manifest %s", path)
+		}
+		if len(b) > 0 {
+			if err = yaml.Unmarshal(b, pm); err != nil {
```

Comment:
I ran `operator-sdk olm-catalog gen-csv --csv-version=0.1.1 --from-version=0.1.0 --update-crds` with an existing package manifest:

```yaml
channels:
- currentCSV: packagemanifest-operator.v0.1.0
  name: stable
packageName: packagemanifest-operator
```

The new manifest file includes `defaultChannel: alpha`, which doesn't seem right:

```yaml
channels:
- currentCSV: packagemanifest-operator.v0.1.0
  name: stable
defaultChannel: alpha
packageName: packagemanifest-operator
```

I think this is happening because we're creating the default manifest and then unmarshalling into it. In the case where the file doesn't set `defaultChannel`, the default `defaultChannel: alpha` (from line 77) sticks around.

Should we declare `pm` without initializing it, and then either create a new default one or unmarshal from the file (but never both) based on the presence of the file?

Lastly, if there's only one channel defined in the package manifest file, should we automatically bump the `currentCSV`?

Reply:
I've updated manifest generation to only use `s.newPackageManifest()` if the file does not exist or is empty so `defaultChannel` is populated correctly`. Also `defaultChannel` will be set if there is only one channel in `channels`, and warnings are emitted if `defaultChannel` is empty or does not exist in `channels`.

I think we should let the user make changes to `currentCSV` in all cases so we don't unexpectedly update something.